### PR TITLE
prevents spurious kind bound errors

### PR DESCRIPTION
**Pull request by xeno-by**
The patch adds a check which makes sure that the trees we're about to
report aren't already erroneous.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @paulp @retronym

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -1003,22 +1003,24 @@ trait Infer extends Checkable {
...
-      } else if (!isWithinBounds(pre, owner, tparams, targs)) {
-        if (!(targs exists (_.isErroneous)) && !(tparams exists (_.isErroneous))) {
+                    tparams: List[Symbol], targs: List[Type], prefix: String): Boolean =
+      if ((targs exists (_.isErroneous)) || (tparams exists (_.isErroneous))) true
+      else {
+        //@M validate variances & bounds of targs wrt variances & bounds of tparams
+        //@M TODO: better place to check this?
+        //@M TODO: errors for getters & setters are reported separately
+        val kindErrors = checkKindBounds(tparams, targs, pre, owner)
+        if (!kindErrors.isEmpty) {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: `if (kindErrors.nonEmpty)`

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -1003,22 +1003,24 @@ trait Infer extends Checkable {
...
+          if (targs contains WildcardType) true
+          else {
+            KindBoundErrors(tree, prefix, targs, tparams, kindErrors)
+            false
+          }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Perhaps:

``` scala
kindErrors match {
  case Nil =>
    def notWithinBounds() = NotWithinBounds(tree, prefix, targs, tparams, Nil)
    isWithinBounds(pre, owner, tparams, targs) || {notWithinBounds(); false}
  case errors =>
    def kindBoundErrors() = KindBoundErrors(tree, prefix, targs, tparams, errors)
    (targs contains WildcardType) || {kindBoundErrors(); false}
}
```

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: That's a nice one! Thank you.

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
LGTM

**Comment by paulp** [comment]
LGTM.

**Pull request merged**